signer/src/network/libp2p/swarm.rs

```diff
@@ -228,19 +314,31 @@ impl<'a> SignerSwarmBuilder<'a> {
     /// Build the [`SignerSwarm`], consuming the builder.
     pub fn build(self) -> Result<SignerSwarm, SignerSwarmError> {
         let keypair: Keypair = (*self.private_key).into();
-        let swarm_config = SignerSwarmConfig {
+        let behavior_config = SignerSwarmConfig {
             enable_mdns: self.enable_mdns,
+            enable_kademlia: self.enable_kademlia,
+            enable_autonat: self.enable_autonat,
             initial_bootstrap_delay: self.initial_bootstrap_delay,
             seed_addresses: self.seed_addrs,
+            num_signers: self.num_signers,
         };
-        let behavior = SignerBehavior::new(keypair.clone(), swarm_config)?;
+        let behavior = SignerBehavior::new(keypair.clone(), behavior_config)?;
+
+        // Protocol configuration
         let noise =
             noise::Config::new(&keypair).map_err(|e| SignerSwarmError::LibP2P(Box::new(e)))?;
         let yamux = yamux::Config::default();
-        let swarm_config = libp2p::swarm::Config::with_tokio_executor()
-            .with_idle_connection_timeout(Duration::from_secs(60));
         let tcp_config = tcp::Config::default().nodelay(true);
 
+        // This is hard-coded to a non-zero u8 value and will never fail.
+        #[allow(clippy::unwrap_in_result)]
+        let dial_concurrency_factor = NonZeroU8::new(4)
+            .ok_or_else(|| SignerSwarmError::Generic("failed to create dial concurrency factor"))?;
```

Comment:
qq: why do we want to use 4 instead of the default 8?

Reply:
It's largely just being conservative, allowing a signer to simultaneously attempt to open connections to enough peers to become well-connected quickly (remember, they only effectively need 1 peer connection to function) while also trying to limit the effects of any potential connection amplification attacks.